hs_access_control/models/resource.py

```diff
@@ -53,24 +53,67 @@ class ResourceAccess(models.Model):
     # workalike queries adapt to old access control system
     #############################################
 
+    @property
+    def __view_users_from_individual(self):
+        return Q(is_active=True,
+                 u2urp__resource=self.resource)
+
+    @property
+    def __view_users_from_group(self):
+        return Q(is_active=True,
+                 u2ugp__group__gaccess__active=True,
+                 u2ugp__group__g2grp__resource=self.resource)
+
+    @property
+    def __view_users_from_community(self):
+        return Q(is_active=True,
+                 u2ugp__group__gaccess__active=True,
+                 u2ugp__group__g2gcp__community__c2gcp__allow_view=True,
+                 u2ugp__group__g2gcp__community__c2gcp__group__gaccess__active=True,
+                 u2ugp__group__g2gcp__community__c2gcp__group__g2grp__resource=self.resource)
+
     @property
     def view_users(self):
         """
         QuerySet of users with view privileges over a resource.
 
         This is a property so that it is a workalike for a prior explicit list.
 
+        This now accounts for group and community privileges
+
         For VIEW, effective privilege = declared privilege, in the sense that all editors have
         VIEW, even if the resource is immutable.
         """
 
-        return User.objects.filter(Q(is_active=True) &
-                                   (Q(u2urp__resource=self.resource,
-                                      u2urp__privilege__lte=PrivilegeCodes.VIEW) |
-                                    Q(u2ugp__group__gaccess__active=True,
-                                      u2ugp__group__g2grp__resource=self.resource,
-                                      u2ugp__group__g2grp__privilege__lte=PrivilegeCodes.VIEW)))\
-                           .distinct()
+        return User.objects.filter(self.__view_users_from_individual |
+                                   self.__view_users_from_group |
+                                   self.__view_users_from_community).distinct()
+
+    @property
+    def __edit_users_from_individual(self):
+        return Q(is_active=True,
+                 u2urp__resource=self.resource,
+                 u2urp__resource__raccess__immutable=False,
+                 u2urp__privilege__lte=PrivilegeCodes.CHANGE)
+
+    @property
+    def __edit_users_from_group(self):
+        return Q(is_active=True,
+                 u2ugp__group__gaccess__active=True,
+                 u2ugp__group__g2grp__resource=self.resource,
+                 u2ugp__group__g2grp__resource__raccess__immutable=False,
+                 u2ugp__group__g2grp__privilege__lte=PrivilegeCodes.CHANGE)
+
+    @property
+    def __edit_users_from_community(self):
+        return Q(
+            is_active=True,
+            u2ugp__group__gaccess__active=True,
+            u2ugp__group__g2gcp__privilege=PrivilegeCodes.CHANGE,
+            u2ugp__group__g2gcp__community__c2gcp__group__gaccess__active=True,
```

Comment:
This filter attribute feels like we are going round a circle. Should this be: `u2ugp__group__g2gcp__group__gaccess__active=True`?

Reply:
Nope. g2gcp maps a group to a community. The community maps via c2gcp to another group. I did not include the clause that excludes loops because I don't care if there are loops in this case.